File: src/models/constraints/compress_vent_constraints.jl
"""
Modeling module
Declare compressor and vented CO2-related constraints.

Pengfei Cheng

1. CO2 compression amount
2. power usage of CO2 compression
3. total CO2 captured
"""

function add_compress_vent_constraints(m)

    x_CO2_compress = m[:x_CO2_compress]
    x_CO2_PCC = m[:x_CO2_PCC]
    x_CO2_DAC = m[:x_CO2_DAC]
    x_power_compress = m[:x_power_compress]
    x_CO2_cap_total = m[:x_CO2_cap_total]
    x_CO2_vent_PCC = m[:x_CO2_vent_PCC]

    # 1. CO2 compression amount
    # integral over all slices within each hour
    @constraint(
        m, eq_compress_CO2[i = set_hour_0], 
        x_CO2_compress[i]
        ==
        x_CO2_PCC[i] 
        + sum(x_CO2_DAC[i, j] for j in set_quarter)
    )

    # ------------------------------------------------------------------------------

    # 2. power usage of CO2 compression
    @constraint(
        m, eq_compress_CO2_power[i = set_hour_0], 
        x_power_compress[i]
        ==
        a_power_compr_PCC * x_CO2_PCC[i]
        + a_power_compr_DAC * sum(x_CO2_DAC[i, j] for j in set_quarter)
    )

    # ------------------------------------------------------------------------------

    # 3. total CO2 captured
    @constraint(
        m, eq_CO2_cap_total[i = set_hour_0], 
        x_CO2_cap_total[i]
        ==
        - x_CO2_vent_PCC[i] 
        + (1 - a_CO2_vent_DAC) * sum(x_CO2_DAC[i, j] for j in set_quarter)
    )
end

File: src/pyomo_model/constraints/compress_vent_constraints.py
"""
Modeling module
Declare compressor and vented CO2-related constraints.

Pengfei Cheng

1. CO2 compression amount
2. power usage of CO2 compression
3. total CO2 captured
"""


from pyomo.environ import Constraint
from ..params import *

def add_compress_vent_constraints(m, set_hour_0, set_scenario):

    # 1. CO2 compression amount
    # integral over all slices within each hour
    def eq_compress_CO2(m, i, s):
        return m.x_CO2_compress[i, s] == m.x_CO2_PCC[i, s] + sum(m.x_CO2_DAC[i, j, s] for j in set_quarter)
    m.eq_compress_CO2 = Constraint(set_hour_0, set_scenario, rule=eq_compress_CO2)

    # --------------------------------------------------------------------------

    # 2. power usage of CO2 compression
    def eq_compress_CO2_power(m, i, s):
        return m.x_power_compress[i, s] == a_power_compr_PCC * m.x_CO2_PCC[i, s] + a_power_compr_DAC * sum(m.x_CO2_DAC[i, j, s] for j in set_quarter)
    m.eq_compress_CO2_power = Constraint(set_hour_0, set_scenario, rule=eq_compress_CO2_power)

    # --------------------------------------------------------------------------

    # 2. power usage of CO2 compression
    def eq_CO2_cap_total(m, i, s):
        return m.x_CO2_cap_total[i, s] == - m.x_CO2_vent_PCC[i, s] + (1 - a_CO2_vent_DAC) * sum(m.x_CO2_DAC[i, j, s] for j in set_quarter)
    m.eq_CO2_cap_total = Constraint(set_hour_0, set_scenario, rule=eq_CO2_cap_total)

    return